File: backend/api/transcription/service.py
from __future__ import annotations

from typing import Any
from uuid import uuid4

from .deepgram_client import request_deepgram_transcription
from .schemas import TranscriptionResponse, Utterance


async def transcribe_audio(
    audio_bytes: bytes,
    filename: str | None,
    content_type: str | None = None,
) -> TranscriptionResponse:
    raw_response = await request_deepgram_transcription(audio_bytes, filename, content_type)
    return normalize_deepgram_response(raw_response)


def normalize_deepgram_response(raw_response: dict[str, Any]) -> TranscriptionResponse:
    metadata = raw_response.get("metadata") or {}
    results = raw_response.get("results") or {}
    alternative = _first_alternative(results)
    provider_request_id = metadata.get("request_id")

    return TranscriptionResponse(
        id=_transcription_id(provider_request_id),
        transcript=str(alternative.get("transcript") or ""),
        confidence=_as_float(alternative.get("confidence")),
        duration=_as_float(metadata.get("duration")),
        utterances=[_normalize_utterance(utterance) for utterance in results.get("utterances") or []],
        provider_request_id=str(provider_request_id) if provider_request_id else None,
        raw_provider_response=raw_response,
    )


def _first_alternative(results: dict[str, Any]) -> dict[str, Any]:
    channels = results.get("channels") or []
    if not channels:
        return {}

    alternatives = channels[0].get("alternatives") or []
    if not alternatives:
        return {}

    return alternatives[0] or {}


def _normalize_utterance(utterance: dict[str, Any]) -> Utterance:
    return Utterance(
        id=str(utterance["id"]) if utterance.get("id") is not None else None,
        start=_as_float(utterance.get("start")),
        end=_as_float(utterance.get("end")),
        confidence=_as_float(utterance.get("confidence")),
        channel=_as_int(utterance.get("channel")),
        speaker=_as_int(utterance.get("speaker")),
        transcript=str(utterance.get("transcript") or ""),
        words=list(utterance.get("words") or []),
    )


def _transcription_id(provider_request_id: Any) -> str:
    if provider_request_id:
        return f"tr_{provider_request_id}"
    return f"tr_{uuid4().hex}"


def _as_float(value: Any) -> float | None:
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _as_int(value: Any) -> int | None:
    if value is None:
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None

File: berkapp/app/_lib/transcription/service.ts
import "server-only";

import { requestDeepgramTranscription } from "./deepgram";
import type { DeepgramResponse, TranscriptionResponse, Utterance } from "./types";

export async function transcribeAudio(
  audioBytes: ArrayBuffer,
  filename?: string,
  contentType?: string,
): Promise<TranscriptionResponse> {
  const rawResponse = await requestDeepgramTranscription(audioBytes, filename, contentType);
  return normalizeDeepgramResponse(rawResponse);
}

export function normalizeDeepgramResponse(rawResponse: DeepgramResponse): TranscriptionResponse {
  const metadata = rawResponse.metadata ?? {};
  const results = rawResponse.results ?? {};
  const alternative = firstAlternative(results);
  const providerRequestId =
    metadata.request_id === undefined || metadata.request_id === null ? null : String(metadata.request_id);

  return {
    id: providerRequestId ? `tr_${providerRequestId}` : `tr_${crypto.randomUUID().replaceAll("-", "")}`,
    transcript: String(alternative?.transcript ?? ""),
    confidence: asNumber(alternative?.confidence),
    duration: asNumber(metadata.duration),
    utterances: (results.utterances ?? []).map(normalizeUtterance),
    provider: "deepgram",
    provider_request_id: providerRequestId,
    raw_provider_response: rawResponse,
  };
}

function firstAlternative(results: NonNullable<DeepgramResponse["results"]>) {
  return results.channels?.[0]?.alternatives?.[0] ?? {};
}

function normalizeUtterance(utterance: Record<string, unknown>): Utterance {
  return {
    id: utterance.id === undefined || utterance.id === null ? null : String(utterance.id),
    start: asNumber(utterance.start),
    end: asNumber(utterance.end),
    confidence: asNumber(utterance.confidence),
    channel: asInteger(utterance.channel),
    speaker: asInteger(utterance.speaker),
    transcript: String(utterance.transcript ?? ""),
    words: Array.isArray(utterance.words) ? (utterance.words as Record<string, unknown>[]) : [],
  };
}

function asNumber(value: unknown) {
  if (value === undefined || value === null) {
    return null;
  }

  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function asInteger(value: unknown) {
  const parsed = asNumber(value);
  return parsed === null ? null : Math.trunc(parsed);
}

